### DISCUSSION TITLE
PlatformIO RP2040 and ST7789 without CS PIN Dark/Black Screen

### DISCUSSION THREAD

**Discussion by Hraefns** [Bodmer/TFT_eSPI#3561] [Q&A - General] [answered]
Hi newcomer here, I need help with making this library to work on my RP2040(YD-RP2040) and ST7789(Without CS PIN) using PlatformIO. I can confirm the PIN is correct and the TFT is not faulty because I already tried to do a flashing background test manually without any library.

I tried to set TFT_BACKLIGHT_ON to LOW and it's work because my TFT screen goes complete black, I can see a greyish flash when I press the reset button too, but after that all I see just complete black screen.
EDIT: I see the lights kind of flashing in loop but it's still black (so it's like backlight go darker and go lighter if that make sense) but this only happens if I power up my pico with another pico (because I was trying to debug it), if I power it up directly through USB C I can see a slight flash when I press the reset button but that's all no flashing after that.

I already tried to search the platformIO.ini example and this is what I came up with but still no luck:
```
[env:ydrp2040]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = vccgnd_yd_rp2040
framework = arduino
board_build.core = earlephilhower
board_build.mcu = rp2040
monitor_speed = 115200

lib_deps = 
    LittleFS
    SPI
    bodmer/TFT_eSPI@^2.5.43
    lvgl/lvgl@8.4.0

build_flags =
    -D PICO_CYW43_SUPPORTED=0 ; To prevent undefined reference __wrap_cyw43_cb_process_ethernet error from building earlephilhower's lib
    -D CYW43_PIN_WL_DYNAMIC=0 ; To prevent undefined reference __wrap_cyw43_cb_process_ethernet error from building earlephilhower's lib
    
    ;###############################################################
    ; TFT_eSPI library setting here (no need to edit library files):
    ;###############################################################
    -Os
    -D USER_SETUP_LOADED=1
    -D ST7789_DRIVER=1

    -D TFT_WIDTH=240
    -D TFT_HEIGHT=240

    -D TFT_SPI_PORT=1
    -D TFT_MISO=-1
    -D TFT_MOSI=15
    -D TFT_SCLK=14
    -D TFT_CS=-1
    -D TFT_RST=12
    -D TFT_DC=11
    -D TFT_BL=10
    -D TFT_BACKLIGHT_ON=HIGH

    -D LOAD_GLCD=1
    -D LOAD_FONT2=1
    -D LOAD_FONT4=1
    -D LOAD_FONT6=1
    -D LOAD_FONT7=1
    -D LOAD_FONT8=1
    -D LOAD_GFXFF=1
    -D SMOOTH_FONT=1

    -D RP2040_PIO_SPI=1
    -D RP2040_DMA=1

    -D SPI_FREQUENCY=27000000
    -D SPI_READ_FREQUENCY=5000000
```

Here is the simple code I used to test with TFT_eSPI:
```
#include <Arduino.h>
#include <TFT_eSPI.h>
#include <SPI.h>

TFT_eSPI tft = TFT_eSPI();

int main()
{
    tft.init();
    // tft.setRotation(1);
    tft.fillScreen(TFT_GREEN);

    while (1)
    {
        tft.fillScreen(TFT_BLUE);
        delay(1000);
        tft.fillScreen(TFT_RED);
        delay(1000);
        tft.fillScreen(TFT_GREEN);
        delay(1000);
        tft.fillScreen(TFT_WHITE);
        delay(1000);
    }

    return 0;
}
```

I don't know if the setup is correct since I'm pretty new to PIO and somehow my earlephilhower will failed without having PICO_CYW43_SUPPORTED & CYW43_PIN_WL_DYNAMIC set to 0. Any help will be really appreciated! :)

**Comment by Basitadam** [comment]
Can you give a try with this pin setup:
```
    -D TFT_SPI_PORT=1
    -D TFT_MISO=-1
    -D TFT_MOSI=11
    -D TFT_SCLK=10
    -D TFT_CS=-1
    -D TFT_RST=12
    -D TFT_DC=9
    -D TFT_BL=14
    -D TFT_BACKLIGHT_ON=HIGH
  ```


with reference to RPi official doc:

![image](https://github.com/user-attachments/assets/f25c3544-3c71-45c3-b8b3-644ccf661a82)


and Adafruit doc:

![image](https://github.com/user-attachments/assets/ab012157-2cdb-4248-a2aa-abc6176e3446)

**Reply by Hraefns** [reply]
Thank you very much for the suggestion, I tried it but I still got a black screen with a little flash at beginning, when I did a reset it just do a blink like this:
https://github.com/user-attachments/assets/0a627b4c-73b2-4139-9e69-877d46bb0a8c
I can confirm the TFT screen is working and I use the same pin as you suggested as well to make sure it's not a solder issue on my previous manually coded SPI:
https://github.com/user-attachments/assets/2e953bdd-31ba-4fac-a5aa-3ffcd66311bd

My RP2040 is from VCC-GND YD-RP2040 with this pinout:
![YD-2040-PIN](https://github.com/user-attachments/assets/6d9c9356-e626-475a-8236-6193f0b8000f)
While my TFT is ST7789 without CS PIN:
![image](https://github.com/user-attachments/assets/a734190b-ff55-428b-9c0c-eabc97027b6d)

I modify my platformIO.ini to be simpler as well:
```
[env:ydrp2040]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = vccgnd_yd_rp2040
framework = arduino
board_build.core = earlephilhower
board_build.mcu = rp2040
monitor_speed = 115200
upload_protocol = picoprobe
debug_tool = picoprobe

lib_deps = 
    ; LittleFS
    ; SPI
    bodmer/TFT_eSPI@^2.5.43
    ;lvgl/lvgl@8.4.0

build_flags =
    -D PICO_CYW43_SUPPORTED=0 ; To prevent undefined reference __wrap_cyw43_cb_process_ethernet error from building earlephilhower's lib
    -D CYW43_PIN_WL_DYNAMIC=0 ; To prevent undefined reference __wrap_cyw43_cb_process_ethernet error from building earlephilhower's lib
    -D CORE_DEBUG_LEVEL=ARDUHAL_LOG_LEVEL_DEBUG
    ;###############################################################
    ; TFT_eSPI library setting here (no need to edit library files):
    ;###############################################################
    -D USER_SETUP_LOADED=1

    -D DISABLE_ALL_LIBRARY_WARNINGS=1

    -D ST7789_DRIVER=1

    -D TFT_WIDTH=240
    -D TFT_HEIGHT=240

    -D TFT_RGB_ORDER=TFT_RGB
    -D TFT_SPI_PORT=1
    ; -D TFT_SPI_MODE=SPI_MODE3

    -D TFT_MISO=-1
    -D TFT_MOSI=11
    -D TFT_SCLK=10
    -D TFT_CS=-1
    -D TFT_RST=12
    -D TFT_DC=9
    -D TFT_BL=14
    -D TFT_BACKLIGHT_ON=HIGH

    -D TOUCH_CS=-1

    -D LOAD_GLCD=0
    -D LOAD_FONT2=0
    -D LOAD_FONT4=0
    -D LOAD_FONT6=0
    -D LOAD_FONT7=0
    -D LOAD_FONT8=0
    -D LOAD_GFXFF=0
    -D SMOOTH_FONT=0

    ; -D RP2040_PIO_SPI=1
    ; -D RP2040_DMA=1

    -D SPI_FREQUENCY=1000000  ; Start with 1 MHz
    -D SPI_READ_FREQUENCY=1000000 
```

I run out of idea how to make it works, I tried to search the same issue but their solutions is not working on me

**Reply by Basitadam** [reply]
There are two other ST7789 divers. Can you give another go with using these one by one:

```-D ST7789_DRIVER_2=1```

and

```-D ST7789_DRIVER_3=1```

**Reply by Hraefns** [reply]
I tried those 2 but I got an error regarding macros was not declared in this scope:
TFT_CASET, TFT_PASET, TFT_RAMWR, TFT_INVON, TFT_INVOFF
I'm not sure what value should put in those macro.

I did try to use ST7789_2_DRIVER which does not give me any error but still not working unfortunately.

**Reply by Basitadam** [reply]
Can you check if you have  "ST7789_3_Defines.h", "ST7789_3_Init.h" and "ST7789_3_Rotation.h" files in your "TFT_eSPI\TFT_Drivers" folder

**Reply by Hraefns** [reply]
Sure, I did a check just now and I don't have those 3 files but I do have these: ST7789_2_Defines.h, ST7789_2_Init.h, and ST7789_2_Rotation.h which used by the ST7789_2_DRIVER.

I think the reason why I got the macros error because there is a check on the TFT_eSPI.h for a driver but I don't see ST7789_DRIVER_2 & ST7789_DRIVER_3 and that create an error since I don't have it defined causing the needed macros absence which I assume should be loaded from those 3 files you mentioned, is that a custom one by any chance? 

Thank you

**Reply by Basitadam** [reply]
Probably. I can send those files but that means I modified TFT_eSPI.cpp and TFT_eSPI.h files,too. You may want to keep those original. If not, I can send those, too.

Actually, you run the display without a library. Can you send your initialization code? I can compare it with the library. If the problem is not with the init, we can check other things

**Reply by Hraefns** [reply]
If that's okay with you, I'm interested to take a look at your modified TFT_eSPI.cpp &  TFT_eSPI.h to learn more about this stuff.

And this is [the code](https://github.com/Hraefns/ST7789-Test/blob/main/src/main.cpp) I used to test the screen with. Thank you very much for helping me.

**Reply by Basitadam** [reply]
Sure. I am sending three _3 driver files, modified TFT_eSPI.cpp & TFT_eSPI.h, and another modified ST7789_init.h file via link.

I checked your working code and modified ST7789_init.h to match yours. Can you check it first? Your setup is very short and simple. Maybe the library is writing some wrong values to your display registers. The only difference with new ST7789_init and yours is color mode. Yours is using rgb888 while library uses rgb565.

If modified ST7789_init.h works but colors are in wrong order, just add this to your PlatformIO setup : ``` -D TFT_RGB_ORDER=0 ```

Here is the link for files:  https://www.filemail.com/d/jxgrquskonpqmmf

**Reply by Hraefns** [reply]
It still doesn't work unfortunately, I'm not sure why. I updated it into like this:
ST7789_init.h
```
#ifndef INIT_SEQUENCE_3
{
  writecommand(ST7789_SWRESET);   // Sleep out
  delay(150);

  writecommand(ST7789_SLPOUT);   // Sleep out
  delay(500);

  writecommand(ST7789_COLMOD);
  writedata(0x55);

  writecommand(ST7789_MADCTL);
  writedata(TFT_MAD_COLOR_ORDER);

  writecommand(ST7789_INVON);
  delay(10);
///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

  end_tft_write();
  delay(120);
  begin_tft_write();

  writecommand(ST7789_DISPON);    //Display on
  delay(120);

#ifdef TFT_BL
  // Turn on the back-light LED
  digitalWrite(TFT_BL, HIGH);
  pinMode(TFT_BL, OUTPUT);
#endif
}
```
And I can confirm that it run the lines because I debug it. I tried to debug the tft.init() to my knowledge I think everything working fine, but I did have doubt in the TFT_eSPI.cpp there is this line 649 that call spi.begin(), which seems to be initialize an SPI peripheral and using CPOL0 for it's begin-end transaction as it call the base implementation from SPI.cpp (Setting it to CPOL 0 could make the screen black which I learnt from manual code the ST7789 but since it end a transaction there maybe it should be safe?) but then I tried to make my own custom function to initialize the SPI:
TFT_eSPI.cpp 
```
/***************************************************************************************
** Function name:           customSpiBegin
***************************************************************************************/
void TFT_eSPI::customSpiBegin(void)
{
    #if defined (TFT_MISO)
    if (TFT_MISO >= 0) {
      gpio_set_function(TFT_MISO, GPIO_FUNC_SPI);
    }
    #endif

    #if defined (TFT_CS) && !defined(RP2040_PIO_INTERFACE)
    if (TFT_CS >= 0) {
      gpio_set_function(TFT_CS, GPIO_FUNC_SPI);
    }
    #endif

    gpio_set_function(TFT_SCLK, GPIO_FUNC_SPI);
    gpio_set_function(TFT_MOSI, GPIO_FUNC_SPI);

    // Give a default config in case user doesn't use beginTransaction
    spi.beginTransaction(SPISettings(SPI_FREQUENCY, MSBFIRST, TFT_SPI_MODE));
    spi.endTransaction();
}
```

But still no luck, this is really weird and I got no idea what could go wrong :(

**Reply by Basitadam** [reply]
Do you have a logic analyzer? If not you can make one using RP2040. There is a ready to use firmware available
Would be best to check the signals at this point.

**Reply by Hraefns** [reply]
Sure, I just got one. I will try to compare between without lib, standard lib, and modified lib (setup). I will get back to you later and thank you for the suggestion

**Reply by Hraefns** [reply]
> Do you have a logic analyzer? If not you can make one using RP2040. There is a ready to use firmware available Would be best to check the signals at this point.

I already done comparing the logic analyzer and yes SPI TFT_eSPI data is different from what it should be:
![image](https://github.com/user-attachments/assets/f8d292d2-177e-4884-9f95-dbaf3ffa801b)

Without Lib the data it's send match the init. I will try to take a look deeper why it have such differences,I tried to set everything the same using:
- 1Mhz
- SPI_MODE_3 (CPOL1, CPHA1)
- Same Pin number
- Set as SPI1
- Same setting on the logic analyzer with SPI_MODE_3 (CPOL1, CPHA1) and 2Mhz scan

**Reply by Basitadam** [reply]
That's it. I wish we did this before.
Can you send me the analyzer log file. Guessing you took it via Logic 2.x.x
We should have read 0x01-0x11-0x3a-0x55-0x36-0x00-0x21-0x29

**Reply by Hraefns** [reply]
Yeah I wish too, I didn't think about the logic analyzer either since I'm not really familiar yet with it, but know I understand how important it is and will try to dig down with it.

Yes, you're correct. I'm using Logic 2.4.14. Here are the files, one with lib and one without lib:
[Initialization Step.zip](https://github.com/user-attachments/files/17919280/Initialization.Step.zip)
And here is the modified [init file](https://drive.google.com/file/d/1zFpqnoywgSCxpvAfBW-C1CNvo5lFqJel/view?usp=sharing) to match with my manual TFT code

**Reply by Basitadam** [reply]
After the first MOSI byte, SCK returns to LOW. Shouldn't. 


![image](https://github.com/user-attachments/assets/f4026a40-ec38-45ed-a2d7-38c9c31be915)





That's why TFT controller (as analyzer) thinks next 8 rising SCK as the next byte (0x08)

![image](https://github.com/user-attachments/assets/9bff6456-9462-41db-b6c8-d835c2431ad5)


without the library, SCK polarity is correct:

![image](https://github.com/user-attachments/assets/0b259c3e-0e85-44cf-a685-b476822bc142)



Find out why SCK is defaulting to LOW level and you are done. SCK should stay HIGH between MOSI datas. Maybe the library is not using hardware SPI for some reason ... etc

... By the way good thinking with lowering the SCK frequency to the half of the analyzer sample frequency. I should have mentioned beforehand

**Reply by Hraefns** [reply]
Thnank you very much for taking your time :D, I really appreciate it. I will try to dig down deeper now, once I find any clues and hopefully an answer I will report here, so if someone faced a similar issue they can find an info here

**Reply by Basitadam** [reply]
Maybe not use these at the same time? Try each alone


```
    -D TFT_SPI_PORT=1
    -D RP2040_PIO_SPI=1
```

and not use DMA until everything works OK:

```
;    -D RP2040_DMA=1
```

**Reply by Hraefns** [reply]
Sure, thanks for the input. I will retested it again since, actually I commented it out for the last couple test. Right know I'm trying to not use spi.begin() which comes from the framework-aruinopico which seems at first glance to use the deafult CPOL 0 & CPHA 0  but I will try to confirm it, hopefully with custom SPI init it works

**Reply by Basitadam** [reply]
I am confident that you will figure it out in couple of hours

**Reply by Hraefns** [reply]
IT'S WORKING NOW!!! Thank you so much so I ended up modifying the TFT_eSPI code I will share it with you tomorrow since it's already late here. But it's working yeah!!!

**Reply by Hraefns** [reply]
I will need to check a few things, maybe there is a macro( and a reason why it's coded that way) so that I don't need to modify the library itself since my knowledge is still green with TFT and stuff. So I need to be careful

**Reply by Basitadam** [reply]
Nice work. It didn't even take an hour

**Comment by Hraefns** [comment, marked as answer]
Okay, I don't find any macro that could be used to fix this but. The fix simply just to modify the **TFT_eSPI_RP2040.h**  just find the macro SET_BUS_READ_MODE and change it into this:
`#define SET_BUS_READ_MODE  spi_set_format(SPI_X,  8, (spi_cpol_t)(TFT_SPI_MODE >> 1), (spi_cpha_t)(TFT_SPI_MODE & 0x1), SPI_MSB_FIRST)`

because if not at the end of each spi transaction (writecommand) it will pull the clock line to SPI MODE 0 (CPOL=0, CPHA=0).

@Basitadam Thank you so much for taking your time to help me, if it wasn't for your help I will still pulling my hairs trying to figure out why the lib doesn't work

**Reply by Basitadam** [reply]
@Bodmer is currently working on RPi Pico AFAIK. Maybe he can take a look at it.
You are a fine engineer. Now you have a responsibility (you should choose to accept it) to monitor the the page and try to help the other Pico users

**Reply by Hraefns** [reply]
Sure, I will try to keep an update to the Pico since I'm planning to use this library alongside the LVGL and will try to help others, just like you help me these past few days :D

**Comment by TobiasVanDyk** [comment]
Maybe this will also help from a few days ago: https://github.com/Bodmer/TFT_eSPI/issues/3476
Yes, reverting to 2.5.34 works ok. I had a look at the difference between 2.5.34 and 2.5.43 (using WinMerge) to see exactly what has changed. And not much, mostly comment updates. But there is a change to this line in TFT_eSPI_RP2040.h:

2.5.34
#define SET_BUS_READ_MODE  // spi_set_format(SPI_X,  8, (spi_cpol_t)0, (spi_cpha_t)0, SPI_MSB_FIRST)

2.5.43
#define SET_BUS_READ_MODE  spi_set_format(SPI_X,  8, (spi_cpol_t)0, (spi_cpha_t)0, SPI_MSB_FIRST)

The spi_set_format function has been reinstated. So I updated to 2.5.42 again and commented out that function and now it's working.

Not sure why this is an issue really, it looks like it should be ok using that function. I don't know enough to investigate further. Hoping others can figure it out from here. :)

Cheers, Ian

**Reply by Hraefns** [reply]
Thank you very much for the info, Tobias. I will definitely try to check that version.
